pinot-core/src/main/java/org/apache/pinot/core/plan/PinotExplainedRelNode.java

```diff
@@ -0,0 +1,146 @@
+ * these RelNodes to the server (using {@code org.apache.pinot.query.planner.plannode.PlanNode} as intermediate format).
+ *
+ * Servers then extract leaf nodes from the stage, trying to execute as much as possible in a leaf operator. In that
+ * leaf operator is where nodes are compiled into single-stage physical operators
+ * (aka {@link org.apache.pinot.core.operator.BaseOperator}), which include important information like whether indexes
+ * are being used or not.
+ *
+ * {@link PinotExplainedRelNode} is a way to represent these single-stage operators in Calcite.
+ * <b>They are not meant to be executed</b>.
+ * Instead, when physical explain is required, the broker ask for the final plan to each server. They return
+ * a PlanNode and the broker then converts these PlanNodes into a PinotExplainedRelNode. The logical plan is then
+ * modified to substitute the nodes that were converted into single-stage physical plans with the corresponding
+ * PinotExplainedRelNode.
+ */
+public class PinotExplainedRelNode extends AbstractRelNode {
+
+  /**
+   * The name of this node, whose role is like a title in the explain plan.
+   */
+  private final String _type;
+  private final Map<String, Plan.ExplainNode.AttributeValue> _attributes;
+  private final List<RelNode> _inputs;
+  private final DataSchema _dataSchema;
+
+  public PinotExplainedRelNode(RelOptCluster cluster, RelTraitSet traitSet, String type,
+      Map<String, Plan.ExplainNode.AttributeValue> attributes, DataSchema dataSchema, RelNode input) {
+    this(cluster, traitSet, type, attributes, dataSchema, Lists.newArrayList(input));
+  }
+
+  public PinotExplainedRelNode(RelOptCluster cluster, RelTraitSet traitSet, String type,
+      Map<String, Plan.ExplainNode.AttributeValue> attributes, DataSchema dataSchema, List<? extends RelNode> inputs) {
+    super(cluster, traitSet);
+    _type = type;
+    _attributes = attributes;
+    _inputs = new ArrayList<>(inputs);
+    _dataSchema = dataSchema;
+  }
+
+  @Override
+  protected RelDataType deriveRowType() {
+    int size = _dataSchema.size();
+    List<RelDataTypeField> fields = new ArrayList<>(size);
+    RelDataTypeFactory typeFactory = getCluster().getTypeFactory();
+    for (int i = 0; i < size; i++) {
+      String columnName = _dataSchema.getColumnName(i);
+
+      DataSchema.ColumnDataType columnDataType = _dataSchema.getColumnDataType(i);
+      RelDataType type = columnDataType.toType(typeFactory);
+
+      fields.add(new RelDataTypeFieldImpl(columnName, i, type));
+    }
+    return new RelRecordType(StructKind.FULLY_QUALIFIED, fields, false);
+  }
+
+  @Override
+  public List<RelNode> getInputs() {
+    return Collections.unmodifiableList(_inputs);
+  }
+
+  @Override
+  public void replaceInput(int ordinalInParent, RelNode p) {
+    _inputs.set(ordinalInParent, p);
+  }
+
+  @Override
+  public String getRelTypeName() {
+    return _type;
+  }
```

Comment:
We need to override this method to because the result is name that is presented in the explain plan. If we don't override it, the explain plan of all these nodes would be `PinotExplainedRelNode` for all cases.